Implement basic layout & behaviour of new mod select screen

bdach opened the pull request:
Continuation of #16917.

Note that - as in title - this is not intended to be a full replacement yet. There is still two-three pulls of work to be done before that; that work includes adding back some extension points that are needed to replace the old overlays, porting over existing test coverage, adding some safeties in mod incompatibility rules, etc. Treat this as a visual appearance / component interaction pass.

https://user-images.githubusercontent.com/20418176/161443186-2c6ed09a-7bd8-4192-86de-86bdce9505dd.mp4

All of the motion design is mostly improvised and up for debate, as is the implementation of scrolls.

bdach committed:
Implement new mod select screen

bdach committed:
Add testing for customisation panel show/hide logic

bdach committed:
Fix propagation of selected mods to columns

bdach committed:
Add testing for mod bindable state propagation

bdach committed:
Add testing for different rulesets

bdach committed:
Improve robustness of test code

peppy committed:
Fix overlapping variable usage

peppy committed:
Adjust pop in/out transitions

peppy committed:
Adjust customisation transition slightly

peppy commented on osu.Game/Overlays/Mods/ModSelectScreen.cs:
```diff
@@ -0,0 +1,357 @@
+
+namespace osu.Game.Overlays.Mods
+{
+    public class ModSelectScreen : OsuFocusedOverlayContainer
```
Naming is a touch weird. A screen which is actually an overlay. Was this mostly to avoid conflicts with the existing class?

bdach replied:
Correct, this was to avoid `ModSelectOverlayV2` or similar atrocities. I figured it was fine to call this a "screen" since it's more... "screeny" than the old overlay, given that it takes up the entire screen (bar toolbar, maybe). Open to suggestions for better names, or to moving these names back to `...Overlay` when this entire ordeal is done.

peppy committed:
Remove unused fields

peppy commented:
I'm currently working on the transitions further, but it's proving to be a bit of a pain due to dark elements overlapping lighter elements in a location I want to perform a fade. Will take another look tomorrow with a fresh set of eyes.

smoogipoo commented:
Just a few things from a quick play:
- I mentioned this long long ago when the design was still WIP, but the whole double scroll thing is really weird.
- The mod customisations should probably appear quicker, and also suffers from the same double scroll weirdness :(

peppy commented:
The bi-directional scroll isn't going anywhere. We might want to do something similar to settings to allow clicking half-offscreen columns and have them come into view to ease navigation.

Another potential improvement would be to have the vertical scrolling only work from drag operations (mouse wheel only controls horizontal).

smoogipoo commented:
My brain explodes any time I have to do horizontal scroll with my vertical mouse wheel. Can't tell you off the top of my head whether a scroll downwards scrolls left-wards or right-wards.

I feel very strongly about this, but I guess we'll never meet eye-to-eye on this and I'll just have to wait until people complain for you and flyte to realise this is a bad UX.

peppy commented:
It's a UX common to many applications (web brosers, apple app store, many top apps). People should be pretty used to it.

Interesting point about "which direction scrolling down scrolls". I didn't think this is a point of contention (right is always down, across every OS/platform).

smoogipoo commented:
Give me an example where an app has left-right scrolling mechanisms on desktops? The only thing I can point to off-hand is apple.com product pages that have fullscreen layouts completely changing when you scroll, but I wouldn't call that left-right.

huoyaoyuan commented:
Some UX opinion at player's side:
The animations (especially open/close) should be fast. Mod selection is a common operation during gameplay, especially for experienced players and they react really fast.

Moltanicaa commented:
I'm sorry to get into such dialogues, but I have a suggestion for this - can you make the customization screen in the form of an opening list under a customizable mod? For example, I want to customize DT - I click on dt, click on the "conditional"customization button (which can be placed on the right side of the DT button) and the mods column expands vertically (only one column), after which the customization options are shown to me.

peppy commented:
Here's a proposal for an initial improvement to scroll behaviour:

```diff
diff --git a/osu.Game/Overlays/Mods/ModColumn.cs b/osu.Game/Overlays/Mods/ModColumn.cs
index 46a562ac98..eea1188303 100644
--- a/osu.Game/Overlays/Mods/ModColumn.cs
+++ b/osu.Game/Overlays/Mods/ModColumn.cs
@@ -151,7 +151,7 @@ public ModColumn(ModType modType, bool allowBulkSelection, Key[]? toggleKeys = n
                                     },
                                     new Drawable[]
                                     {
-                                        new OsuScrollContainer
+                                        new LocalScrollContainer
                                         {
                                             RelativeSizeAxes = Axes.Both,
                                             ScrollbarOverlapsContent = false,
@@ -194,6 +194,29 @@ public ModColumn(ModType modType, bool allowBulkSelection, Key[]? toggleKeys = n
             }
         }
 
+        public class LocalScrollContainer : OsuScrollContainer
+        {
+            [Resolved]
+            private OsuGameBase game { get; set; } = null!;
+
+            protected override bool OnScroll(ScrollEvent e)
+            {
+                if (!game.ScreenSpaceDrawQuad.Contains(ScreenSpaceDrawQuad.TopRight))
+                    return false;
+
+                if (!game.ScreenSpaceDrawQuad.Contains(ScreenSpaceDrawQuad.BottomLeft))
+                    return false;
+
+                if (e.ScrollDelta.Y > 0 && Target <= 0)
+                    return false;
+
+                if (e.ScrollDelta.Y < 0 && IsScrolledToEnd())
+                    return false;
+
+                return base.OnScroll(e);
+            }
+        }
+
         private void createHeaderText()
         {
             IEnumerable<string> headerTextWords = ModType.Humanize(LetterCasing.Title).Split(' ');
```

bdach committed:
Always play pop in sequence

bdach committed:
Fix weird test step

bdach commented:
Not sure about the above diff, feels pretty wonky at times:

https://user-images.githubusercontent.com/20418176/161607603-a38609da-6353-471c-b774-65bd290e99ef.mp4

I was going to suggest a variant of this which disallows vertical scroll if there is nothing to scroll in the column, and blocks horizontal scroll if there is something to scroll in the column:

https://user-images.githubusercontent.com/20418176/161607756-d21b0517-368b-4b75-a473-170996143a36.mp4

which is similar to what the beatmap cards do, but still has some degree of wonk. Diff for this:

<details>

<summary>this comment is too long already but here's a diff</summary>

```diff
diff --git a/osu.Game/Overlays/Mods/ModColumn.cs b/osu.Game/Overlays/Mods/ModColumn.cs
index 46a562ac98..94d712a3ac 100644
--- a/osu.Game/Overlays/Mods/ModColumn.cs
+++ b/osu.Game/Overlays/Mods/ModColumn.cs
@@ -1,6 +1,8 @@
 // Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
+#nullable enable
+
 using System;
 using System.Collections.Generic;
 using System.Linq;
@@ -16,6 +18,7 @@
 using osu.Framework.Graphics.Shapes;
 using osu.Framework.Graphics.Sprites;
 using osu.Framework.Input.Events;
+using osu.Framework.Utils;
 using osu.Game.Graphics;
 using osu.Game.Graphics.Containers;
 using osu.Game.Graphics.UserInterface;
@@ -25,8 +28,6 @@
 using osuTK.Graphics;
 using osuTK.Input;
 
-#nullable enable
-
 namespace osu.Game.Overlays.Mods
 {
     public class ModColumn : CompositeDrawable
@@ -151,7 +152,7 @@ public ModColumn(ModType modType, bool allowBulkSelection, Key[]? toggleKeys = n
                                     },
                                     new Drawable[]
                                     {
-                                        new OsuScrollContainer
+                                        new LocalScrollContainer
                                         {
                                             RelativeSizeAxes = Axes.Both,
                                             ScrollbarOverlapsContent = false,
@@ -194,6 +195,26 @@ public ModColumn(ModType modType, bool allowBulkSelection, Key[]? toggleKeys = n
             }
         }
 
+        public class LocalScrollContainer : OsuScrollContainer
+        {
+            [Resolved]
+            private OsuGameBase game { get; set; } = null!;
+
+            protected override bool OnScroll(ScrollEvent e)
+            {
+                if (!game.ScreenSpaceDrawQuad.Contains(ScreenSpaceDrawQuad.TopRight))
+                    return false;
+
+                if (!game.ScreenSpaceDrawQuad.Contains(ScreenSpaceDrawQuad.BottomLeft))
+                    return false;
+
+                if (Precision.AlmostEquals(ScrollableExtent, 0))
+                    return false;
+
+                return base.OnScroll(e);
+            }
+        }
+
         private void createHeaderText()
         {
             IEnumerable<string> headerTextWords = ModType.Humanize(LetterCasing.Title).Split(' ');
```

</details>

bdach commented on osu.Game/Overlays/Mods/ModSelectScreen.cs:
```diff
@@ -0,0 +1,343 @@
+
+            customisationVisible.BindValueChanged(_ => updateCustomisationVisualState(), true);
+
+            FinishTransforms(true);
```
I don't actually recall why I needed this so I just made the transition fire on enter even if the overlay starts already in visible state.

peppy commented:
If the "wonky" part is the scroll-past-end behaviour, maybe disabling that for this view is best. I think the advantage of the method I proposed is that by scrolling you will go through all mods, from start to finish, across all columns, without having to move the cursor.

peppy commented:
@smoogipoo @bdach able to give this version a try? i think it feels pretty good. i've duplicated mod display to give more columns more mods

```diff
diff --git a/osu.Game/Overlays/Mods/ModColumn.cs b/osu.Game/Overlays/Mods/ModColumn.cs
index 46a562ac98..2088a834d5 100644
--- a/osu.Game/Overlays/Mods/ModColumn.cs
+++ b/osu.Game/Overlays/Mods/ModColumn.cs
@@ -151,7 +151,7 @@ public ModColumn(ModType modType, bool allowBulkSelection, Key[]? toggleKeys = n
                                     },
                                     new Drawable[]
                                     {
-                                        new OsuScrollContainer
+                                        new LocalScrollContainer
                                         {
                                             RelativeSizeAxes = Axes.Both,
                                             ScrollbarOverlapsContent = false,
@@ -194,6 +194,59 @@ public ModColumn(ModType modType, bool allowBulkSelection, Key[]? toggleKeys = n
             }
         }
 
+        private class LocalScrollContainer : OsuScrollContainer
+        {
+            private OsuScrollContainer? parentScrollContainer;
+
+            protected override void LoadComplete()
+            {
+                base.LoadComplete();
+
+                parentScrollContainer = this.findClosestParent<OsuScrollContainer>();
+
+                ClampExtension = 100;
+            }
+
+            protected override bool OnScroll(ScrollEvent e)
+            {
+                if (parentScrollContainer?.ScreenSpaceDrawQuad.Contains(ScreenSpaceDrawQuad.TopRight) == false
+                    || parentScrollContainer?.ScreenSpaceDrawQuad.Contains(ScreenSpaceDrawQuad.BottomLeft) == false)
+                {
+                    // If not on screen, handle scroll but also allow parent to scrol at the same time.
+                    base.OnScroll(e);
+                    return false;
+                }
+
+                if (e.ScrollDelta.Y > 0 && Target <= 0)
+                {
+                    base.OnScroll(e);
+                    return false;
+                }
+
+                if (e.ScrollDelta.Y < 0 && IsScrolledToEnd())
+                {
+                    base.OnScroll(e);
+                    return false;
+                }
+
+                return base.OnScroll(e);
+            }
+
+            // TODO: remove when https://github.com/ppy/osu-framework/pull/5092 is available.
+            private T? findClosestParent<T>() where T : class, IDrawable
+            {
+                Drawable cursor = this;
+
+                while ((cursor = cursor.Parent) != null)
+                {
+                    if (cursor is T match)
+                        return match;
+                }
+
+                return default;
+            }
+        }
+
         private void createHeaderText()
         {
             IEnumerable<string> headerTextWords = ModType.Humanize(LetterCasing.Title).Split(' ');
@@ -251,7 +304,16 @@ private void updateMods()
             var panels = newMods.Select(mod => new ModPanel(mod)
             {
                 Shear = new Vector2(-ModPanel.SHEAR_X, 0)
-            });
+            }).ToList();
+
+            panels.AddRange(newMods.Select(mod => new ModPanel(mod)
+            {
+                Shear = new Vector2(-ModPanel.SHEAR_X, 0)
+            }));
+            panels.AddRange(newMods.Select(mod => new ModPanel(mod)
+            {
+                Shear = new Vector2(-ModPanel.SHEAR_X, 0)
+            }));
 
             Task? loadTask;
 
diff --git a/osu.Game/Overlays/Mods/ModSelectScreen.cs b/osu.Game/Overlays/Mods/ModSelectScreen.cs
index 7915fb755c..42794a5505 100644
--- a/osu.Game/Overlays/Mods/ModSelectScreen.cs
+++ b/osu.Game/Overlays/Mods/ModSelectScreen.cs
@@ -100,13 +100,13 @@ private void load()
                                     {
                                         RelativeSizeAxes = Axes.Both,
                                         RelativePositionAxes = Axes.Both,
-                                        Padding = new MarginPadding { Horizontal = 70 },
                                         Children = new Drawable[]
                                         {
                                             new OsuScrollContainer(Direction.Horizontal)
                                             {
                                                 RelativeSizeAxes = Axes.Both,
                                                 Masking = false,
+                                                ClampExtension = 100,
                                                 ScrollbarOverlapsContent = false,
                                                 Child = columnFlow = new ModColumnContainer
                                                 {

```

https://user-images.githubusercontent.com/191335/161702414-7e3a194f-ce2a-4a32-b8cd-cfa85eecf38d.mp4

peppy committed:
Move scroll view padding to content level

Without doing this, there is a non-masked but also non-interactive area
to the left or right of the view.

peppy committed:
Improve nested scroll behaviour

peppy commented:
I'm going to push the above as I think it's in a pretty good state to iterate off of.

peppy committed:
Simplify and better comment nested scroll conditionals

smoogipoo commented:
Seems weird to me that the inner scroll containers scroll even when they're at their extents. At this point I don't think any input I provide is going to be of use so I'm going to refrain from doing so.

peppy commented:
I was going to propose you test without that added behaviour in the comment but not sure it feels much better. I like the sense of movement.

You can test quickly by removing the `base.` calls from the two conditionals.

peppy committed:
Speed up customisation panel toggle a bit

smoogipoo commented:
To me it feels better without the `base.` calls. Anyway that's the last bit of input I'm giving here.

peppy commented:
Yeah I agree with that. Will make the change.

peppy committed:
Limit `FillFlow` of columns to applicable direction

peppy committed:
Slightly adjust scroll handling and also apply to `ModSettingsContainer`

peppy committed:
Fix multiplier display resetting transformations

peppy committed:
Animate multiplier display

peppy committed:
Animate individual `ModColumn`s during togle of oerlay

peppy committed:
Adjust animation metrics slightly

peppy committed:
Remove left padding to allow left-most column to exist further to the.. left

peppy commented:
I've got the animation roughly in a state I'm happy with. Also made a few other adjustments. Would appreciate a second set of eyes.

peppy committed:
Use newly exposed framework helper function to find closest parent

peppy committed:
Add missing newline

bdach committed:
Merge branch 'master' into mod-overlay/full-screen

bdach commented:
Not sure if I count for the second part of eyes but the latest changes look good to me.

I've also pushed a merge commit here to fix the csproj conflicts.
